### PR TITLE
Timeout HashTable Task

### PR DESCRIPTION
**Pull request by ivaylo91**
TimeoutHashTable Task Ready.

### CODE REVIEW

File: src/addremovethreads/App.java [hunk 1]
```diff
@@ -10,7 +10,7 @@ public static void main(String[] args) throws InterruptedException {
 
         AddElement addElement = new AddElement(listOfElements, "Bai Tosho");
 
-        RemoveElement removeElement = new RemoveElement(listOfElements,"Bai Tosho");
+        RemoveElement removeElement = new RemoveElement(listOfElements, "Bai Tosho");
```
Comment (mgenov) [inline src/addremovethreads/App.java]: RemoveElement is a verb.

File: src/timeouthashtable/TimeoutHashTable.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.Map;
+
+/**
+ * @author Ivaylo Penev(ipenev91@gmail.com)
+ */
+public class TimeoutHashTable<String, Object> {
+
+    private Map<String, TimeoutRemover<String, Object>> timeoutHashTable = new Hashtable<>();
+    private long timeout;
+    TimeoutRemover<String, Object> thread;
```
Comment (mgenov) [inline src/timeouthashtable/TimeoutHashTable.java]: Why the timeout remover doesn't have modifier ?

File: src/timeouthashtable/TimeoutHashTable.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    public TimeoutHashTable(long timeout) {
+        this.timeout = timeout;
+    }
+
+    public void put(String key, Object value) {
+        thread = new TimeoutRemover<>(timeoutHashTable, key, value, timeout);
```
Comment (mgenov) [inline src/timeouthashtable/TimeoutHashTable.java]: What happens if key already exists? Do you start another thread ? 

```java
put("key1", "value1"); 
put("key1", "value2"); // bam ! Starts another thread for the same key which is expensive operation and should be prevented. 
```

File: src/timeouthashtable/TimeoutHashTable.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    public Object get(String key) {
+        if (!timeoutHashTable.containsKey(key)) {
+            return null;
+        } else {
+            return timeoutHashTable.get(key).getValue();
+        }
+    }
+
+    public Object remove(String key) {
+        if (timeoutHashTable.containsKey(key)) {
```
Comment (mgenov) [inline src/timeouthashtable/TimeoutHashTable.java]: What happens with the background thread that is counting for that key?

File: src/timeouthashtable/TimeoutHashTable.java (hunk 3 above)
Comment (ivaylo91) [inline src/timeouthashtable/TimeoutHashTable.java]: If the key already exist, when I put the same key, old will be removed and in hash-table will put the new key with new value. I will not start another thread.

Comment (mgenov) [inline src/timeouthashtable/TimeoutHashTable.java]: Then what thread.start() does?

File: src/timeouthashtable/TimeoutHashTable.java (hunk 2 above)
Comment (ivaylo91) [inline src/timeouthashtable/TimeoutHashTable.java]: Fixed ! Added modifier.

File: src/timeouthashtable/TimeoutHashTable.java (hunk 3 above)
Comment (ivaylo91) [inline src/timeouthashtable/TimeoutHashTable.java]: Fixed! Remove thread.start().

**Comment by mgenov** [comment]
Actually I'm not seeing one of the requirements. When element is accessed, the counting has to start again, cause in your case, there is no difference whether element is accessed or not.

File: src/timeouthashtable/TimeoutHashTable.java (hunk 4 above)
Comment (ivaylo91) [inline src/timeouthashtable/TimeoutHashTable.java]: Fixed!

File: src/timeouthashtable/TimeoutHashTable.java (hunk 3 above)
Comment (ivaylo91) [inline src/timeouthashtable/TimeoutHashTable.java]: Fixed! Add if statement that check key already exist or not, if key do not exist add to the timeouthashtable.

**Comment by mgenov** [comment]
LGTM !

**Pull request closed without merging**